File: app/src/main/java/su/linka/pictures/components/OutputLine.java
package su.linka.pictures.components;

import android.content.Context;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.GridView;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.io.IOException;
import java.util.ArrayList;

import su.linka.pictures.Card;
import su.linka.pictures.OutputAdapter;
import su.linka.pictures.R;
import su.linka.pictures.Set;
import su.linka.pictures.SetManifest;
import su.linka.pictures.TTS;

public class OutputLine extends LinearLayout {

    private final TTS tts;
    private final MediaPlayer mp;
    private Set set;
    private SetManifest manifest;
    private ImageButton backSpaceButton;
    private ImageButton speakButton;
    private ImageButton clearButton;
    private TextView textOutputView;
    private OutputGrid grid;
    private boolean withoutSpace;
    private OutputAdapter adapter;

    private ArrayList<Card> cards = new ArrayList<>();
    private boolean directMode;
    private int currentPlayCard = 0;
    private boolean isPlaying = false;


    public OutputLine(Context context, AttributeSet set){

        super(context, set);
        inflate(context);
        
        tts = new TTS(context);

        mp = new MediaPlayer();

        mp.setAudioStreamType(AudioManager.STREAM_MUSIC);
    }
    private void inflate(Context context) {
        LayoutInflater layoutInflater = (LayoutInflater) context
                .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        layoutInflater.inflate(R.layout.output_line, this, true);
        
        backSpaceButton = findViewById(R.id.backspace_button);
        speakButton = findViewById(R.id.speak_button);
        clearButton = findViewById(R.id.clear_button);;
        textOutputView = findViewById(R.id.output_text);
        grid = findViewById(R.id.output_grid);

        backSpaceButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                backspace();
            }
        });
        clearButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                clear();
            }
        });
        speakButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                speak();
            }
        });
    }

    private void speak() {
        if(withoutSpace){
            tts.speak(getText());
        } else {
            if (isPlaying) {
                stop();
            } else {
                playCards();
            }
        }
    }

    private void playCards() {
        if(isPlaying) return;;

        currentPlayCard = 0;
        isPlaying = true;
        if(cards.size()>0) {
            playCard();
        }
    }

    private void playCard() {
        play(cards.get(currentPlayCard), new OnPlayedListener() {
            @Override
            public void onPlayed() {
                if(!isPlaying) return;
                currentPlayCard++;
                if(currentPlayCard<cards.size()){
                    playCard();
                }
            }
        });
    }

    private void play(Card card, OnPlayedListener onPlayedListener) {

        try {
            mp.setDataSource(set.getAudioFile(card.audioPath).getAbsolutePath());
            mp.prepare();

            mp.setOnCompletionListener(new MediaPlayer.OnCompletionListener() {
                @Override
                public void onCompletion(MediaPlayer mp) {
                    onPlayedListener.onPlayed();
                }
            });
            mp.start();

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void stop(){
        isPlaying=false;
        mp.stop();
    }

    private void clear() {
        cards.clear();
        if(withoutSpace){
            updateText();
        } else {
            grid.clear();
        }
    }

    private void backspace() {
        if(cards.size()==0) return;
        cards.remove(cards.size()-1);
        if(withoutSpace){
            updateText();
        } else {
            grid.backspace();
        }
    }

    public void setSet(Set set) {
        this.set = set;
        manifest =  set.getManifest();
        withoutSpace = manifest.withoutSpace;

        grid.setVisibility(withoutSpace?GONE:VISIBLE);
        textOutputView.setVisibility(withoutSpace?VISIBLE:GONE);

        grid.setSet(set);

    }

    public Set getSet() {
        return set;
    }

    public void addCard(Card newCard) {
        cards.add(newCard);
        if(withoutSpace){
            updateText();
        }else {

            grid.addCard(newCard);
        }
    }

    private void updateText() {
        
        textOutputView.setText(getText());
    }

    private String getText() {
        StringBuilder builder = new StringBuilder();

        for (int i = 0; i < cards.size(); i++) {
            Card card = cards.get(i);

            if(card.cardType==0){
                builder.append(card.title);
            } else if(card.cardType==1){
                builder.append(' ');
            }
        }
        return builder.toString();
    }

    public void setDirectMode(boolean directMode) {
        this.directMode = directMode;
    }

    public boolean getDirectMode() {
        return directMode;
    }

    public static abstract class OnPlayedListener{
        public abstract void onPlayed();
    }
}

File: app/src/main/java/su/linka/pictures/components/OutputLine.kt
package su.linka.pictures.components

import android.app.Activity
import android.content.Context
import android.media.AudioManager
import android.media.MediaPlayer
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.HorizontalScrollView
import android.widget.ImageButton
import android.widget.LinearLayout
import android.widget.TextView
import androidx.annotation.IdRes
import su.linka.pictures.Card
import su.linka.pictures.R
import su.linka.pictures.Set
import su.linka.pictures.SetManifest
import su.linka.pictures.TTS
import java.io.File
import java.io.IOException
import java.util.concurrent.atomic.AtomicBoolean

class OutputLine @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null
) : LinearLayout(context, attrs) {

    private val tts = TTS(context)
    private val mediaPlayer = MediaPlayer().apply {
        @Suppress("DEPRECATION")
        setAudioStreamType(AudioManager.STREAM_MUSIC)
    }

    private var set: Set? = null
    private var manifest: SetManifest? = null
    private val backSpaceButton: ImageButton
    private val speakButton: ImageButton
    private val clearButton: ImageButton
    private val textOutputView: TextView
    private val grid: OutputGrid
    private val scrollView: HorizontalScrollView

    private val cards = mutableListOf<Card>()
    private var withoutSpace: Boolean = false
    private var directMode: Boolean = false
    private var currentPlayCard: Int = 0
    private val isPlaying = AtomicBoolean(false)

    init {
        inflate(context)
    }

    private fun inflate(context: Context) {
        val inflater = LayoutInflater.from(context)
        inflater.inflate(R.layout.output_line, this, true)

        backSpaceButton = requireView(R.id.backspace_button)
        speakButton = requireView(R.id.speak_button)
        clearButton = requireView(R.id.clear_button)
        textOutputView = requireView(R.id.output_text)
        grid = requireView(R.id.output_grid)
        scrollView = requireView(R.id.scroll_grid)

        backSpaceButton.setOnClickListener { backspace() }
        clearButton.setOnClickListener { clear() }
        speakButton.setOnClickListener { speak() }
    }

    fun setSet(set: Set) {
        this.set = set
        manifest = set.getManifest()
        withoutSpace = manifest?.withoutSpace ?: false

        scrollView.visibility = if (withoutSpace) View.GONE else View.VISIBLE
        textOutputView.visibility = if (withoutSpace) View.VISIBLE else View.GONE

        grid.setSet(set, output = true)
    }

    fun getSet(): Set? = set

    fun addCard(newCard: Card) {
        if (directMode) {
            play(newCard, null)
            return
        }
        cards.add(newCard)
        if (withoutSpace) {
            updateText()
        } else {
            grid.addCard(newCard)
            postScrollToEnd()
        }
    }

    fun stop() {
        isPlaying.set(false)
        try {
            if (mediaPlayer.isPlaying) {
                mediaPlayer.stop()
            }
        } catch (_: Exception) {
        }
    }

    fun clear() {
        cards.clear()
        if (withoutSpace) {
            updateText()
        } else {
            grid.clear()
        }
    }

    fun backspace() {
        if (cards.isEmpty()) return
        cards.removeAt(cards.lastIndex)
        if (withoutSpace) {
            updateText()
        } else {
            grid.backspace()
        }
    }

    fun setDirectMode(directMode: Boolean) {
        this.directMode = directMode
    }

    fun release() {
        stop()
        mediaPlayer.release()
        tts.shutdown()
    }

    private fun speak() {
        if (withoutSpace) {
            tts.speak(getText())
        } else {
            if (isPlaying.get()) {
                stop()
            } else {
                playCards()
            }
        }
    }

    private fun playCards() {
        if (!isPlaying.compareAndSet(false, true)) return
        if (cards.isEmpty()) {
            isPlaying.set(false)
            return
        }
        currentPlayCard = 0
        playNextCard()
    }

    private fun playNextCard() {
        if (!isPlaying.get()) return
        val card = cards.getOrNull(currentPlayCard)
        if (card == null) {
            isPlaying.set(false)
            return
        }
        play(card) {
            if (!isPlaying.get()) return@play
            currentPlayCard++
            if (currentPlayCard < cards.size) {
                playNextCard()
            } else {
                isPlaying.set(false)
            }
        }
    }

    private fun play(card: Card, onPlayedListener: (() -> Unit)?) {
        val set = set ?: return
        val audio: File = set.getAudioFile(card.audioPath) ?: run {
            onPlayedListener?.invoke()
            return
        }
        try {
            mediaPlayer.reset()
            mediaPlayer.setDataSource(audio.absolutePath)
            mediaPlayer.setOnCompletionListener {
                onPlayedListener?.invoke()
            }
            mediaPlayer.prepare()
            mediaPlayer.start()
        } catch (error: IOException) {
            error.printStackTrace()
            onPlayedListener?.invoke()
        }
    }

    private fun updateText() {
        textOutputView.text = getText()
    }

    private fun getText(): String {
        val builder = StringBuilder()
        cards.forEach { card ->
            when (card.cardType) {
                0 -> builder.append(card.title)
                1 -> builder.append(' ')
            }
        }
        return builder.toString()
    }

    private fun postScrollToEnd() {
        val activity = context as? Activity
        val runnable = Runnable {
            scrollView.fullScroll(View.FOCUS_RIGHT)
        }
        if (activity != null) {
            activity.runOnUiThread(runnable)
        } else {
            scrollView.post(runnable)
        }
    }

    private fun <T : View> requireView(@IdRes id: Int): T {
        return findViewById<T>(id)
            ?: throw IllegalStateException("View with id $id not found in OutputLine")
    }
}
